x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java

```diff
@@ -301,30 +303,53 @@ public void createJobResultIndex(Job job, ClusterState state, final ActionListen
                                 // so we need to handle that possibility
                                 if (e instanceof ResourceAlreadyExistsException) {
                                     LOGGER.info("Index already exists");
-                                    // Create the alias
-                                    createAliasListener.onResponse(true);
+                                    // Add the term field mappings and alias.  The complication is that the state at the
+                                    // beginning of the operation doesn't have any knowledge of the index, as it's only
+                                    // just been created.  So we need yet another operation to get the mappings for it.
+                                    getLatestIndexMappings(indexName, ActionListener.wrap(
+                                        response -> {
+                                            // Expect one index and one type.  If this is not the case then it means the
+                                            // index has been deleted almost immediately after being created, and this is
+                                            // so unlikely that it's reasonable to fail the whole operation.
+                                            ImmutableOpenMap<String, MappingMetaData> indexMappings =
+                                                response.getMappings().iterator().next().value;
+                                            MappingMetaData typeMappings = indexMappings.iterator().next().value;
+                                            addTermsAndAliases(typeMappings, indexName, termFields, createAliasListener);
```

Comment:
It seems to me that a race condition around mappings still exists, though it is an equally rare one. 
* 3 calls to create the results index
* It gets created with by one of the calls (with those term mappings)
* The other two fail and fall into this clause, pull that single job's mapping, update it with their own fields, and put the whole mapping again
* The first of the two "late" jobs would lose its terms due to the second of the "late" jobs replacing the mapping on with `PutMappingRequest`

Reply:
> * The first of the two "late" jobs would lose its terms due to the second of the "late" jobs replacing the mapping on with `PutMappingRequest`

No, because if you put mappings they always _add_ to the current mappings, they don't replace them.  This is how Elasticsearch mappings have always worked and is why we get away with _only_ putting mappings for the new job terms.  We don't put our core result field mappings except when the index is initially created and they don't get wiped out.

You are correct there is still a race condition, but it's on the number of fields in the mappings.  If you create 3 jobs simultaneously each with 400 new term fields then the one that creates the index will add the first 400.  The other two will see that their 400 plus the 400 already in the index is 800, which is OK, so will both try to put mappings.  And whichever of these requests arrives second will fail because then the index would have too many field mappings: 1200 plus however many core fields we have.  However, this has always been a bug as it could also happen in the code path where the index already exists.  The end result of this is that you'd get an unfriendly error message about mapping limits instead of a friendly one telling you you need to use a custom results index for the job, so it's not _that_ bad considering the tiny chance of it happening.

Reply:
@droberts195 I read
```
// Put the whole mapping, not just the term fields, otherwise we'll wipe the _meta section of the mapping
try (XContentBuilder termFieldsMapping = ElasticsearchMappings.resultsMapping(mappingType, termFields)) {
```
incorrectly. I keep forgetting that an `UPDATE` to a mapping is done via a PUT and not a POST...

The number of fields is still a race condition, but only when it comes to failing early, which is acceptable.